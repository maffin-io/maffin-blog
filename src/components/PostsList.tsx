import { DateTime } from 'luxon';
import Link from 'next/link';

import Tag from '@/components/Tag';
import type { Post } from '@/app/api/getPosts';

type PostsListProps = {
  posts: Post[],
};

export default function PostsList({ posts }: PostsListProps): JSX.Element {
  return (
    <ul className="divide-y divide-gray-200 dark:divide-gray-700">
      {posts.map((post) => (
        <li key={post.slug} className="py-12">
          <article>
            <div className="space-y-2 xl:grid xl:grid-cols-4 xl:items-baseline xl:space-y-0">
              <dl>
                <dt className="sr-only">Published on</dt>
                <dd className="text-base font-medium leading-6 text-gray-500 dark:text-gray-400">
                  <time dateTime={post.date.toISODate() || undefined}>
                    {post.date.toLocaleString(DateTime.DATE_MED)}
                  </time>
                </dd>
              </dl>
              <div className="space-y-5 xl:col-span-3">
                <div className="space-y-6">
                  <div>
                    <h2 className="text-2xl font-bold leading-8 tracking-tight">
                      <Link
                        href={`/posts/${post.slug}`}
                        className="text-gray-900 dark:text-gray-100"
                      >
                        {post.title}
                      </Link>
                    </h2>
                    <div className="flex flex-wrap">
                      {post.tags.map((tag: string) => (
                        <Tag key={tag} text={tag} />
                      ))}
                    </div>
                  </div>
                  <div className="prose max-w-none text-gray-500 dark:text-gray-400">
                    {post.summary}
                  </div>
                </div>
                <div className="text-base font-medium leading-6">
                  <Link
                    href={`/posts/${post.slug}`}
                    className="text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                    aria-label={`Read "${post.title}"`}
                  >
                    Read more (
                    {post.reading_time}
                    ) &rarr;
                  </Link>
                </div>
              </div>
            </div>
          </article>
        </li>
      ))}
    </ul>
  );
}
