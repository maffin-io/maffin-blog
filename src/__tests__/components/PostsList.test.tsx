import { DateTime } from 'luxon';

import PostsList from '@/components/PostsList';
import { render } from '@testing-library/react';

describe('PostsList', () => {
  it('renders all posts as expected', () => {
    const posts = [
      {
        content: '<div>Content</div>',
        slug: 'my-blog-post-1',
        summary: 'Blog post summary TODO',
        tags: ['label1', 'label2'],
        reading_time: '1 min',
        title: 'My Blog Post 1',
        date: DateTime.fromISO('2023-01-01'),
        author: {
          name: 'Name',
          avatar: 'https://avatar.image',
        },
      },
      {
        content: '<div>Content</div>',
        slug: 'my-blog-post-2',
        summary: 'Blog post summary TODO',
        tags: ['label2'],
        reading_time: '1 min',
        title: 'My Blog Post 2',
        date: DateTime.fromISO('2023-01-01'),
        author: {
          name: 'Name',
          avatar: 'https://avatar.image',
        },
      },
    ];

    const { container } = render(<PostsList posts={posts} />);

    expect(container).toMatchSnapshot();
  });
});
