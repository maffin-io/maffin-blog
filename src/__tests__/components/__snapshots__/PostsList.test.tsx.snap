// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList renders all posts as expected 1`] = `
<div>
  <ul
    class="divide-y divide-gray-200 dark:divide-gray-700"
  >
    <li
      class="py-12"
    >
      <article>
        <div
          class="space-y-2 xl:grid xl:grid-cols-4 xl:items-baseline xl:space-y-0"
        >
          <dl>
            <dt
              class="sr-only"
            >
              Published on
            </dt>
            <dd
              class="text-base font-medium leading-6 text-gray-500 dark:text-gray-400"
            >
              <time
                datetime="2023-01-01"
              >
                Jan 1, 2023
              </time>
            </dd>
          </dl>
          <div
            class="space-y-5 xl:col-span-3"
          >
            <div
              class="space-y-6"
            >
              <div>
                <h2
                  class="text-2xl font-bold leading-8 tracking-tight"
                >
                  <a
                    class="text-gray-900 dark:text-gray-100"
                    href="/posts/my-blog-post-1"
                  >
                    My Blog Post 1
                  </a>
                </h2>
                <div
                  class="flex flex-wrap"
                >
                  <a
                    href="/tags/label1"
                  >
                    <span
                      class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                    >
                      #
                      label1
                    </span>
                  </a>
                  <a
                    href="/tags/label2"
                  >
                    <span
                      class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                    >
                      #
                      label2
                    </span>
                  </a>
                </div>
              </div>
              <div
                class="prose max-w-none text-gray-500 dark:text-gray-400"
              >
                Blog post summary TODO
              </div>
            </div>
            <div
              class="text-base font-medium leading-6"
            >
              <a
                aria-label="Read "My Blog Post 1""
                class="text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                href="/posts/my-blog-post-1"
              >
                Read more (
                1 min
                ) →
              </a>
            </div>
          </div>
        </div>
      </article>
    </li>
    <li
      class="py-12"
    >
      <article>
        <div
          class="space-y-2 xl:grid xl:grid-cols-4 xl:items-baseline xl:space-y-0"
        >
          <dl>
            <dt
              class="sr-only"
            >
              Published on
            </dt>
            <dd
              class="text-base font-medium leading-6 text-gray-500 dark:text-gray-400"
            >
              <time
                datetime="2023-01-01"
              >
                Jan 1, 2023
              </time>
            </dd>
          </dl>
          <div
            class="space-y-5 xl:col-span-3"
          >
            <div
              class="space-y-6"
            >
              <div>
                <h2
                  class="text-2xl font-bold leading-8 tracking-tight"
                >
                  <a
                    class="text-gray-900 dark:text-gray-100"
                    href="/posts/my-blog-post-2"
                  >
                    My Blog Post 2
                  </a>
                </h2>
                <div
                  class="flex flex-wrap"
                >
                  <a
                    href="/tags/label2"
                  >
                    <span
                      class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                    >
                      #
                      label2
                    </span>
                  </a>
                </div>
              </div>
              <div
                class="prose max-w-none text-gray-500 dark:text-gray-400"
              >
                Blog post summary TODO
              </div>
            </div>
            <div
              class="text-base font-medium leading-6"
            >
              <a
                aria-label="Read "My Blog Post 2""
                class="text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                href="/posts/my-blog-post-2"
              >
                Read more (
                1 min
                ) →
              </a>
            </div>
          </div>
        </div>
      </article>
    </li>
  </ul>
</div>
`;
